x-pack/metricbeat/module/gcp/metrics/dataproc/metadata.go

```diff
@@ -0,0 +1,208 @@
+// you may not use this file except in compliance with the Elastic License.
+
+package dataproc
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	monitoringpb "cloud.google.com/go/monitoring/apiv3/v2/monitoringpb"
+	"google.golang.org/api/dataproc/v1"
+	"google.golang.org/api/option"
+
+	"github.com/elastic/beats/v7/x-pack/metricbeat/module/gcp"
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+// NewMetadataService returns the specific Metadata service for a GCP Dataproc cluster
+func NewMetadataService(projectID string, regions []string, organizationID, organizationName string, projectName string, collectUserLabels bool, opt ...option.ClientOption) (gcp.MetadataService, error) {
+	return &metadataCollector{
+		projectID:         projectID,
+		projectName:       projectName,
+		organizationID:    organizationID,
+		organizationName:  organizationName,
+		regions:           regions,
+		collectUserLabels: collectUserLabels,
+		opt:               opt,
+		clusters:          make(map[string]*dataproc.Cluster),
+		logger:            logp.NewLogger("metrics-dataproc"),
+	}, nil
+}
+
+// dataprocMetadata is an object to store data in between the extraction and the writing in the destination (to uncouple
+// reading and writing in the same method)
+type dataprocMetadata struct {
+	region      string
+	clusterID   string
+	clusterName string
+	machineType string
+
+	User     map[string]string
+	Metadata map[string]string
+	Metrics  interface{}
+	System   interface{}
+}
+
+type metadataCollector struct {
+	projectID         string
+	projectName       string
+	organizationID    string
+	organizationName  string
+	regions           []string
+	collectUserLabels bool
+	opt               []option.ClientOption
+	// NOTE: clusters holds data used for all metrics collected in a given period
+	// this avoids calling the remote endpoint for each metric, which would take a long time overall
+	clusters map[string]*dataproc.Cluster
+	logger   *logp.Logger
+}
+
+// Metadata implements googlecloud.MetadataCollector to the known set of labels from a Dataproc TimeSeries single point of data.
+func (s *metadataCollector) Metadata(ctx context.Context, resp *monitoringpb.TimeSeries) (gcp.MetadataCollectorData, error) {
+	stackdriverLabels := gcp.NewStackdriverMetadataServiceForTimeSeries(resp, s.organizationID, s.organizationName, s.projectName)
+
+	metadataCollectorData, err := stackdriverLabels.Metadata(ctx, resp)
+	if err != nil {
+		return gcp.MetadataCollectorData{}, err
+	}
+
+	if resp.Resource != nil && resp.Resource.Labels != nil {
+		_, _ = metadataCollectorData.ECS.Put(gcp.ECSCloudInstanceIDKey, resp.Resource.Labels["cluster_uuid"])
+	}
+
+	_, _ = metadataCollectorData.ECS.Put(gcp.ECSCloudInstanceNameKey, resp.Resource.Labels["cluster_name"])
+
+	if s.collectUserLabels {
+		metadata, err := s.instanceMetadata(ctx, s.instanceID(resp), s.instanceRegion(resp))
+		if err != nil {
+			return gcp.MetadataCollectorData{}, err
```

Comment:
do we want to return the object we populated with cluster UUID and cluster name above here (`metadataCollectorData`) ?

Reply:
Yes, they fit in the `cloud.instance.id` and `cloud.instance.name` ECS fields. Are you suggesting setting them only if `collectUserLabels` is enabled?

Reply:
no i'm wondering why we don't return the object with the attributes set. instead we return an empty struct.

Reply:
oh i see, yeah you are right we could still return `metadataCollectorData` even if `instanceMetadata` errors out.